### PR TITLE
core(unsized-images): skip images that are zero-sized

### PR DESCRIPTION
**Pull request by paulirish**
running kohls i saw these in the unsized-images results..

image

of course anything that's zero width & height isn't part of the layout (equivalent to display:none) but we hadn't really been checking that.

the boundingRect we have ends up being ideal for this calculation and includes if a parent is hidden or w/e.

sweet.

### CODE REVIEW

File: lighthouse-core/audits/unsized-images.js [hunk 1]
```diff
@@ -91,14 +91,22 @@ class UnsizedImages extends Audit {
...
       const isFixedImage =
         image.cssComputedPosition === 'fixed' || image.cssComputedPosition === 'absolute';
+      if (isFixedImage) continue;
 
-      if (isFixedImage || UnsizedImages.isSizedImage(image)) continue;
-      const url = URL.elideDataURI(image.src);
+      // Zero-sized images won't cause layout shift. We approve!
+      const boundingRect = image.node.boundingRect;
+      // Either the image was set to 0,0 size or a parent is not displayed.
+      const isNotDisplayed = boundingRect.width === 0 && boundingRect.height === 0;
```
Comment (connorjclark) [inline lighthouse-core/audits/unsized-images.js:102]: `||` ?

Comment (patrickhulce) [reply lighthouse-core/audits/unsized-images.js:102]: I don't think that solves the CLS issue this audit is about though. If an item has 0 width but shoves text down because it has height, we're still in trouble, no?

**Review by patrickhulce — COMMENTED** [review]
this policy is great, but there's something worse going wrong if these are being flagged because they have an explicit CSS size of 0 already and should be ignored for that reason.

I'm concerned about the deeper issue here :/

If we don't have time to look into it now, maybe we file an issue but save this repro with what to uncomment?

File: lighthouse-core/audits/unsized-images.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/audits/unsized-images.js:102]: oof, i didn't realize elements of zero width can take up height/any space.. weird but ok

Comment (patrickhulce) [reply lighthouse-core/audits/unsized-images.js:102]: layout is fun :)

**Comment by paulirish** [comment]
> this policy is great, but there's something worse going wrong if these are being flagged because they have an explicit CSS size of 0 already and should be ignored for that reason.

you mean explicit **_HTML_** size of 0, right?

image


> I'm concerned about the deeper issue here :/

yeah...

here's the function which handles zero's oddly..


https://github.com/GoogleChrome/lighthouse/blob/b292ae06ec9666d2a26c6dd44f835fa4f79bf1bf/lighthouse-core/audits/unsized-images.js#L42-L52


the non-negative bit [comes from the spec](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width).

the spec also says "If the two attributes are both zero, it indicates that the element is not intended for the user (e.g. it might be a part of a service to count page views)." which is nice.


but, I don't really follow why 0 as a width/height attribute size would not be a _valid sized_ attr value.

----

soo.... i'll update this fn to not reject cases where attr value is "0"

**Comment by patrickhulce** [comment]
> you mean explicit HTML size of 0, right?

No, I was actually talking about the inline CSS `style="width:0px; height:0px;"` but thanks for finding an even better bug on my mistake! 😆 

> soo.... i'll update this fn to not reject cases where attr value is "0"

SG 👍

**Comment by paulirish** [comment]
> > you mean explicit HTML size of 0, right?
> 
> No, I was actually talking about the inline CSS `style="width:0px; height:0px;"` but thanks for finding an even better bug on my mistake! 😆

ah true. yeah.


well i've fixed the HTML one just now. 😛 

small detail about CSS width/height 0.. so with this PR (and whatever change we're about to make to the gatherer (re #11289)) .. it's possible for images where 0/0 is set in inline styles... will make it through to the results.. because we'd only see the 0/0 from fetchSourceRules and soon we're not guaranteed to be doing that on all our images.  but... hey i think that's okay.

File: lighthouse-core/audits/unsized-images.js [hunk 2]
```diff
@@ -40,15 +40,23 @@ class UnsizedImages extends Audit {
...
+   * @url https://html.spec.whatwg.org/multipage/embedded-content-other.html#dimension-attributes
+   * @param {string} attrValue
    * @return {boolean}
    */
-  static isValidAttr(attr) {
-    const NON_NEGATIVE_INT_REGEX = /^\d+$/;
-    const ZERO_REGEX = /^0+$/;
-    return NON_NEGATIVE_INT_REGEX.test(attr) && !ZERO_REGEX.test(attr);
+  static isValidAttr(attrValue) {
+    // First, superweird edge case of using the positive-sign. The the spec _sorta_ says it's valid...
```
Comment (patrickhulce) [inline lighthouse-core/audits/unsized-images.js:49]:
```suggestion
    // First, superweird edge case of using the positive-sign. The spec _sorta_ says it's valid...
```

File: lighthouse-core/audits/unsized-images.js [hunk 3]
```diff
@@ -40,15 +40,23 @@ class UnsizedImages extends Audit {
...
+    // https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-integers
+    //   > Otherwise, if the character is … (+): Advance position to the next character.
+    //   > (The "+" is ignored, but it is not conforming.)
+    // lol.  Chrome (at least) doesn't ignore but rejects this as a non-conforming value.
```
Comment (patrickhulce) [inline lighthouse-core/audits/unsized-images.js:53]:
```suggestion
    // lol.  Irrelevant though b/c Chrome (at least) doesn't ignore. It rejects this as a non-conforming value.
```

or any other phrasing that emphasizes a little more why we should ignore what the spec says :)

File: lighthouse-core/test/audits/unsized-images-test.js [hunk 4]
```diff
@@ -234,6 +234,38 @@ describe('Sized images audit', () => {
...
+          boundingRect: {
+            width: 0,
+            height: 0,
+          },
+        },
+      });
+      expect(result.score).toEqual(1);
+    });
+
+    it('passes when an image has is unsized, but its parent is not displayed..', async () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/unsized-images-test.js:254]:
```suggestion
    it('passes when an image has is unsized, but its parent is not displayed', async () => {
```

were those intentional?

File: lighthouse-core/test/audits/unsized-images-test.js [hunk 5]
```diff
@@ -401,19 +433,19 @@ describe('Size attribute validity check', () => {
...
-  it('fails on non-numeric characters', () => {
+  it('handles non-numeric edgecases', () => {
     expect(UnsizedImagesAudit.isValidAttr('zero')).toEqual(false);
-    expect(UnsizedImagesAudit.isValidAttr('1002$')).toEqual(false);
+    expect(UnsizedImagesAudit.isValidAttr('1002$')).toEqual(true); // interpretted as 1002
     expect(UnsizedImagesAudit.isValidAttr('s-5')).toEqual(false);
-    expect(UnsizedImagesAudit.isValidAttr('3,000')).toEqual(false);
-    expect(UnsizedImagesAudit.isValidAttr('100.0')).toEqual(false);
-    expect(UnsizedImagesAudit.isValidAttr('2/3')).toEqual(false);
+    expect(UnsizedImagesAudit.isValidAttr('3,000')).toEqual(true); // interpretted as 3
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/unsized-images-test.js:440]: perhaps the subject of a different future audit :)

**Review by patrickhulce — APPROVED** [review]

File: lighthouse-core/audits/unsized-images.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-core/audits/unsized-images.js:53]: good call thx.
